omnigibson/simulator.py

```diff
@@ -602,27 +614,57 @@ def _on_contact(self, contact_headers, contact_data):
         For each of the pair of objects in each contact, we invoke the on_contact function for each of its states
         that subclass ContactSubscribedStateMixin. These states update based on contact events.
         """
-        if gm.ENABLE_OBJECT_STATES:
-            combos = set()
+        if gm.ENABLE_OBJECT_STATES and self._objects_require_contact_callback:
             headers = defaultdict(list)
             for contact_header in contact_headers:
                 actor0 = str(PhysicsSchemaTools.intToSdfPath(contact_header.actor0))
                 actor1 = str(PhysicsSchemaTools.intToSdfPath(contact_header.actor1))
                 # actor0/1 are prim paths for links that are in contact. Find the corresponding objects.
                 actor0_obj = self._scene.object_registry("prim_path", "/".join(actor0.split("/")[:-1]))
                 actor1_obj = self._scene.object_registry("prim_path", "/".join(actor1.split("/")[:-1]))
-                if actor0_obj is None or actor1_obj is None or not actor0_obj.initialized or not actor1_obj.initialized:
+                # If any of the objects cannot be found, skip
+                if actor0_obj is None or actor1_obj is None:
                     continue
-                headers[tuple(sorted((actor0_obj, actor1_obj), key=lambda x:x.uuid))].append(contact_header)
-
-            for (actor0_obj, actor1_obj) in combos:
+                # If any of the objects is not initialized, skip
+                if not actor0_obj.initialized or not actor1_obj.initialized:
+                    continue
+                # If any of the objects is not stateful, skip
                 if not isinstance(actor0_obj, StatefulObject) or not isinstance(actor1_obj, StatefulObject):
                     continue
+                # If any of the objects doesn't have states that require on_contact callbacks, skip
+                if len(actor0_obj.states.keys() & self.object_state_types_on_contact) == 0 or len(actor1_obj.states.keys() & self.object_state_types_on_contact) == 0:
+                    continue
+                headers[tuple(sorted((actor0_obj, actor1_obj), key=lambda x: x.name))].append(contact_header)
```

Comment:
oops, i guess we will resolve this merge conflict later?

Reply:
will do this after you merge your PR.